### PR TITLE
Add `format_args_capture` feature

### PR DESCRIPTION
**Pull request by davidhewitt**
This is the initial implementation PR for [RFC 2795](https://github.com/rust-lang/rfcs/pull/2795).

Note that, as dicussed in the tracking issue (#67984), the feature gate has been called `format_args_capture`.

Next up I guess I need to add documentation for this feature. I've not written any docs before for rustc / std so I would appreciate suggestions on where I should add docs.

### CODE REVIEW

**Comment by davidhewitt** [comment]
Looks like was a formatting failure, have force-pushed with tidied formatting.

**Comment by poliorcetics** [comment]
You can use https://rustc-dev-guide.rust-lang.org/implementing_new_features.html and the two pages following it to read more about fully implementing a feature.

File: src/librustc_builtin_macros/format.rs [hunk 1]
```diff
@@ -498,10 +501,59 @@ impl<'a, 'b> Context<'a, 'b> {
...
-                        let mut err = self.ecx.struct_span_err(sp, &msg[..]);
-                        err.emit();
+                        let capture_feature_enabled = self
+                            .ecx
+                            .ecfg
+                            .features
+                            .map_or(false, |features| features.format_args_capture);
+
+                        // For the moment capturing variables from format strings expanded from
+                        // literals is disabled (see RFC #2795)
```
Comment (izik1) [inline src/librustc_builtin_macros/format.rs:511]: Would it be correct to assume that either this comment or the below line is wrong (this comment implies `!self.is_literal`)

File: src/librustc_builtin_macros/format.rs [hunk 2]
```diff
@@ -498,10 +501,59 @@ impl<'a, 'b> Context<'a, 'b> {
...
+                                self.fmtsp
+                            };
+                            let mut err = self.ecx.struct_span_err(sp, &msg[..]);
+
+                            if capture_feature_enabled && !self.is_literal {
+                                err.note(&format!(
+                                    "did you intend to capture a variable `{}` from \
+                                     the surrounding scope?",
+                                    name
+                                ));
```
Comment (estebank) [inline src/librustc_builtin_macros/format.rs:538]: We have moved away from the wording "did you mean..." towards a style that doesn't ask questions. You could unify both `note`s and reword it to to "`format_args!()` cannot capture `{}` from the surrounding scope when the format string is expanded from a macro".

File: src/librustc_builtin_macros/format.rs [hunk 3]
```diff
@@ -498,10 +501,59 @@ impl<'a, 'b> Context<'a, 'b> {
...
+                            } else if self.ecx.parse_sess().unstable_features.is_nightly_build() {
+                                err.note(&format!(
+                                    "did you intend to capture a variable `{}` from \
+                                     the surrounding scope?",
+                                    name
+                                ));
+                                err.help(
+                                    "add `#![feature(format_args_capture)]` to the crate \
+                                     attributes to enable",
+                                );
```
Comment (estebank) [inline src/librustc_builtin_macros/format.rs:552]: "if you meant to capture `{}` from the surrounding scope, add `#![feature(format_args_capture)]` to the crate"

File: src/librustc_builtin_macros/format.rs (hunk 1 above)
Comment (davidhewitt) [reply src/librustc_builtin_macros/format.rs:511]: Yes, thanks for catching that!

**Comment by davidhewitt** [comment]
Thanks for the feedback; I've adjusted messaging based on it 👍

**Comment by davidhewitt** [comment]
I've added a commit which (I think) will add a page documenting this feature to the unstable book.

Please let me know any other locations where it would be good to document this feature.

(Maybe in [`std::fmt` documentation](https://doc.rust-lang.org/std/fmt/index.html)? Or should that wait until stabilization PR?)

**Comment by varkor** [comment]
Sorry, will try to get to this soon (it's been a busy week)!

File: src/librustc_builtin_macros/format.rs [hunk 4]
```diff
@@ -498,10 +501,55 @@ impl<'a, 'b> Context<'a, 'b> {
...
+                            let mut err = self.ecx.struct_span_err(sp, &msg[..]);
+
+                            if capture_feature_enabled && !self.is_literal {
+                                err.note(&format!(
+                                    "did you intend to capture a variable `{}` from \
+                                     the surrounding scope?",
+                                    name
+                                ));
+                                err.note(
+                                    "for hygiene reasons format_args! cannot capture variables \
```
Comment (varkor) [inline src/librustc_builtin_macros/format.rs:540]:
```suggestion
                                    "`format_args!` cannot capture variables \
```
I think the reason will be clear, but if you prefer to have one, I think "hygiene reasons" is a little informal. Perhaps "to avoid ambiguity" is clearer.

**Review by varkor — COMMENTED** [review]
This looks good to me. I just have one comment.

File: src/librustc_builtin_macros/format.rs (hunk 4 above)
Comment (davidhewitt) [reply src/librustc_builtin_macros/format.rs]: Thanks agreed. I prefer give a reason, so I've gone for "to avoid ambiguity".

File: src/librustc_builtin_macros/format.rs [hunk 5]
```diff
@@ -498,10 +501,55 @@ impl<'a, 'b> Context<'a, 'b> {
...
+                            let mut err = self.ecx.struct_span_err(sp, &msg[..]);
+
+                            if capture_feature_enabled && !self.is_literal {
+                                err.note(&format!(
+                                    "did you intend to capture a variable `{}` from \
+                                     the surrounding scope?",
+                                    name
+                                ));
+                                err.note(
+                                    "to avoid ambiguity format_args! cannot capture variables \
```
Comment (varkor) [inline src/librustc_builtin_macros/format.rs:540]:
```suggestion
                                    "to avoid ambiguity, `format_args!` cannot capture variables \
```

Comment (davidhewitt) [reply src/librustc_builtin_macros/format.rs]: Thanks, I agree that's better

**Comment by davidhewitt** [comment]
I'll fix this evening the test failure caused by our change to the diagnostic wording

**Comment by varkor** [comment]
This looks good to me; thanks for the pull request!

@bors r+

**Comment by Manishearth** [comment]
@bors r-

https://github.com/rust-lang/rust/pull/73944#issuecomment-652730009

**Comment by davidhewitt** [comment]
Looks like the tests for `panic!` formatting failed on wasm because there is no unwind there (it just triggers unreachable). 

Is it fine to just add `#[cfg(not(target_arch = "wasm32"))]` to those tests?

**Comment by varkor** [comment]
You can add a comment to the relevant tests like this:
https://github.com/rust-lang/rust/blob/f8b796b10b3282b02a6bb6f5b5be9dc8dc0dbf94/src/test/codegen/global_asm_include.rs#L37-L39

**Comment by davidhewitt** [comment]
Thanks, I've pushed a tweak which adds `ignore-wasm32` and `ignore-wasm64`, I think that should be enough.

Really a more elegant solution would be to add a `#[cfg(panic = "unwind")]` flag to skip just the offending tests on targets where panic is an abort.

I couldn't find such a `cfg` flag - do you think it would be worth me opening a PR to implement it?

**Comment by varkor** [comment]
Thanks. @bors r+

> I couldn't find such a cfg flag - do you think it would be worth me opening a PR to implement it?

I'm not sure what use-cases this would have other than for tests. Maybe you could open up an issue if you can think of others, though?

**Comment by davidhewitt** [comment]
> I'm not sure what use cases this would have other than for tests. Maybe you could open up an issue if you can think of others, though?

So far I've only seen it useful for tests as you say. I've seen this same issue crop up in pyO3.

I'll leave it for now.
